### PR TITLE
Commit existing framework internal documentation (#6)

### PR DESCRIPTION
**Pull request by tsjackson-noaa**
**Description**

This PR adds documentation of the framework internals to the documentation website. "Internals" refers to the framework code in `/src`; this information is only relevant for the package's maintainers, not POD developers or end users, and this is stated in a warning banner.

Changes consist of 

1. Expanded and corrected code docstrings, which are processed by Sphinx-autodoc and now included in a section of the documentation website (Internal code documentation/Module Index). 
2. Tweaks to Sphinx conf.py to improve the quality of autodoc-generated output.
3. A higher-level walkthrough of some of the more complex aspects of the code: i) how runtime configuration is set from the CLI and configuration files and passed to the rest of the code; ii) the "data layer" which performs data search and preprocessing based on POD's configuration; iii) commonly used utility modules. 
4. Changes to the HTML theme in `doc/conf.py` and via a custom CSS file in `doc/_static/custom.css`, in order to make the documentation website more legible (especially for the autodoc-generated pages). This is subjective, though, and these changes may be dropped or modified.
5. Corrections to syntax errors in .rst documentation files for existing PODs.

The current PR is a rebase of tsjackson-noaa/MDTF-diagnostics/feature/docs-internals-2. All conflicts with `main` have been resolved.

Associated issues #6 ; #317 

**How Has This Been Tested?**
HTML and latex documentation build locally under Sphinx 3.1.2 (version specified in the Conda environment for development). No changes have been made to executable code.

**Checklist:**
- [X] I have reviewed my own code to ensure that if follows the [POD development guidelines](https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx/dev_guidelines.html)
- [X] My branch is up-to-date with the NOAA-GFDL develop branch, and all merge conflicts are resolved
- [ ] The scripts are written in Python 3.7 or above (preferred; required if funded by a CPO grant), NCL, or R
- [ ] All of my scripts are in the diagnostics/[POD short name] subdirectory, and include a main_driver script, template html, and settings.jsonc file
- [X] I have made corresponding changes to the documentation in the POD's doc/ subdirectory
- [ ] If applicable, I've added a .yml file to src/conda, and my environment builds with `conda_env_setup.sh` 
- [ ] I have added any necessary data to input_data/obs_data/[pod short name] and/or input_data/model/[pod short name]
- [ ] My code is portable; it uses [MDTF environment variables](https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx/ref_envvars.html), and does not contain hard-coded file or directory paths
- [ ] I have provided the code to generate digested data files from raw data files
- [ ] Each digested data file generated by the script contains numerical data (no figures), and is 3 GB or less in size
- [X] The repository contains no extra test scripts or data files

### CODE REVIEW

File: src/preprocessor.py [hunk 1]
```diff
@@ -565,11 +672,11 @@ def __init__(self, data_mgr, pod):
...
         self.parser = self._XarrayParserClass(data_mgr, pod)
         # initialize PreprocessorFunctionBase objects
-        self.functions = [cls_(data_mgr, pod) for cls_ in self._functions]
+        self.functions = [cls_(data_mgr, pod) for cls_ in self.functions]
 
     @property
-    def _functions(self):
-        """Determine which preprocessor functions are applicable to the current
+    def functions(self):
+        """Determine which PreprocessorFunctions are applicable to the current
```
Comment (wrongkindofdoctor) [inline src/preprocessor.py:679]: Removing the `_` from the functions definition and reference in the list results in a Nonetype for the pod.preprocessor in the setup_pod routine, which causes the CI to fail. Reverting it should solve the issue.

File: src/data_sources.py [hunk 2]
```diff
@@ -214,7 +214,7 @@ class MetadataRewritePreprocessor(preprocessor.DaskMultiFilePreprocessor):
     :class:`ExplicitFileDataSourceConfigEntry` objects in the \config_by_id
     attribute of :class:`ExplicitFileDataSource`.
     """
-    _file_preproc_functions = []
+    file_preproc_functions = []
```
Comment (wrongkindofdoctor) [inline src/data_sources.py:217]: Could you submit modifications to the variable definitions in a separate PR? The file_preproc_functions changes don't appear to cause problems like the _functions change did, but it would be better to limit this PR to docs changes.

File: src/preprocessor.py [hunk 3]
```diff
@@ -918,4 +1056,4 @@ class DefaultPreprocessor(DaskMultiFilePreprocessor):
     """Implementation class for :class:`MDTFPreprocessorBase` for the general
     use case. Includes all implemented functionality and handles multi-file data.
     """
-    _file_preproc_functions = []
+    file_preproc_functions = []
```
Comment (wrongkindofdoctor) [inline src/preprocessor.py:1059]: Move variable name mods to separate PR

**Review by wrongkindofdoctor — CHANGES_REQUESTED** [review]
@tsjackson-noaa Look who's emerged from the ether! Thanks for cleaning up the docs.

File: src/preprocessor.py (hunk 1 above)
Comment (tsjackson-noaa) [reply src/preprocessor.py]: Thanks for diagnosing the source of the error so quickly, @wrongkindofdoctor! Reverted in [130e994](https://github.com/NOAA-GFDL/MDTF-diagnostics/pull/372/commits/130e994fcaf3a78a0ec4188e2a375e4c0f5cc5d9).

Because the attributes are private, their docstrings won't be included in the Sphinx autodoc docs (which I'm guessing is why I changed them in the first place), but that's a minor difference.

File: src/data_sources.py (hunk 2 above)
Comment (tsjackson-noaa) [reply src/data_sources.py]: Resolved in [130e994](https://github.com/NOAA-GFDL/MDTF-diagnostics/pull/372/commits/130e994fcaf3a78a0ec4188e2a375e4c0f5cc5d9), see https://github.com/NOAA-GFDL/MDTF-diagnostics/pull/372#discussion_r897516962

File: src/preprocessor.py [hunk 4]
```diff
@@ -845,23 +978,28 @@ def read_dataset(self, var):
...
-    preprocessing model data provided as one or several netcdf files per
-    variable.
+    """A Preprocessor class that uses xarray's dask support to
+    preprocess model data provided as one or multiple netcdf files per
+    variable, using xarray `open_mfdataset()
+    <https://xarray.pydata.org/en/stable/generated/xarray.open_mfdataset.html>`__.
     """
-    _file_preproc_functions = util.abstract_attribute()
+    file_preproc_functions = util.abstract_attribute()
+    """List of PreprocessorFunctions to be executed on a per-file basis as the
```
[the remark this answers is not in the document]
Comment (tsjackson-noaa) [reply src/preprocessor.py]: Resolved in [130e994](https://github.com/NOAA-GFDL/MDTF-diagnostics/pull/372/commits/130e994fcaf3a78a0ec4188e2a375e4c0f5cc5d9), see https://github.com/NOAA-GFDL/MDTF-diagnostics/pull/372#discussion_r897516962

File: src/preprocessor.py [hunk 5]
```diff
@@ -845,23 +978,28 @@ def read_dataset(self, var):
...
+    multi-file Dataset is being loaded, rather than afterwards as part of the
+    :meth:`process`. Note that such functions will not be able to rely on the
+    metadata cleaning done by xr_parser."""
 
     def __init__(self, data_mgr, pod):
         super(DaskMultiFilePreprocessor, self).__init__(data_mgr, pod)
         # initialize PreprocessorFunctionBase objects
         self.file_preproc_functions = \
-            [cls_(data_mgr, pod) for cls_ in self._file_preproc_functions]
+            [cls_(data_mgr, pod) for cls_ in self.file_preproc_functions]
```
[the remark this answers is not in the document]
Comment (tsjackson-noaa) [reply src/preprocessor.py]: Resolved in [130e994](https://github.com/NOAA-GFDL/MDTF-diagnostics/pull/372/commits/130e994fcaf3a78a0ec4188e2a375e4c0f5cc5d9), see https://github.com/NOAA-GFDL/MDTF-diagnostics/pull/372#discussion_r897516962

File: src/preprocessor.py (hunk 3 above)
Comment (tsjackson-noaa) [reply src/preprocessor.py]: Resolved in [130e994](https://github.com/NOAA-GFDL/MDTF-diagnostics/pull/372/commits/130e994fcaf3a78a0ec4188e2a375e4c0f5cc5d9), see https://github.com/NOAA-GFDL/MDTF-diagnostics/pull/372#discussion_r897516962

**Review by wrongkindofdoctor — APPROVED** [review]
